projects/driver_controls/src/stm32f0xx/ads1015.c

```diff
@@ -0,0 +1,171 @@
+#include "ads1015.h"
+// All the important data about ADS1015 is stored and changed in ADS1015Storage storage by different
+// functions. Storage has the channel_enable bitset to know which channels are enabled. It alse
+// holds the current channel that's being converted, the readings, channel callbacks, etc.
+// Ads1015_init writes the desired settings to registers and inits the GPIO pin (ALRT/RDY). (This
+// pin is asserted when a conversion result is ready). Ads1015_init registers prv_interrupt_handler
+// as the callback for this pin. Once the interrupt is raised, it means that a conversion has been
+// completed. prv_interrupt_handler reads the conversion values from the registers and stores them
+// in storage, and then switches the channel to the next enabled one.
+// The user could access conversion data by the provided read functions. These
+// functions get their data from storage and not directly by reading the registers. Also look at
+// ads1015_def.h for more information on macros and how registers are setup.
+// Reading and writing to registers is done through I2C interface. Look at section 8.5.3 of the
+// datasheet for the timing diagrams.
+#include <status.h>
+#include <stdio.h>
+#include <string.h>
+#include "ads1015_def.h"
+#include "gpio_it.h"
+
+// Updates the channel_enable bitset when a channel is enabled/disabled.
+static void prv_mark_channel_enabled(Ads1015Channel channel, bool enable, uint8_t *channel_enable) {
+  if (enable) {
+    *channel_enable |= 1 << channel;
+  } else {
+    *channel_enable &= ~(1 << channel);
+  }
+}
+
+// Writes to register given upper and lower bytes.
+static StatusCode prv_setup_register(Ads1015Storage *storage, uint8_t reg, uint8_t msb,
+                                     uint8_t lsb) {
+  uint8_t ads1015_setup_register[] = { reg, msb, lsb };
+  return i2c_write(storage->i2c_port, storage->i2c_addr + ADS1015_I2C_BASE_ADDRESS,
+                   ads1015_setup_register, SIZEOF_ARRAY(ads1015_setup_register));
+}
+
+// Reads the register and stores the value in the given array.
+static StatusCode prv_read_register(I2CPort i2c_port, Ads1015Address i2c_addr, uint8_t reg,
+                                    uint8_t *rx_data, size_t rx_len) {
+  status_ok_or_return(i2c_write(i2c_port, i2c_addr + ADS1015_I2C_BASE_ADDRESS, &reg, sizeof(reg)));
```

Comment:
Honestly, I feel like you could just calculate the correct address once and store it.

Reply:
Yes that's so much better!